filters/include/pcl/filters/impl/farthest_point_sampling.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception, Inc.
+ *
+ *  All rights reserved.
+ */
+
+#pragma once
+
+#include <pcl/common/geometry.h>
+#include <pcl/filters/farthest_point_sampling.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <algorithm>
+#include <limits>
+#include <random>
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (Indices &indices)
+{
+  const std::size_t size = input_->size();
+  //if requested number of point is equal to the point cloud size, copy original cloud
+  if (sample_size_ == size)
+  {
+    indices = *indices_;
+    removed_indices_->clear ();
+    return;
+  }
+  //check if requested number of points is greater than the point cloud size
+  if (sample_size_ > size)
+  {
+    PCL_THROW_EXCEPTION (BadArgumentException,
+                        "Requested number of points is greater than point cloud size!");
+  }
+
+  std::vector<float> distances_to_selected_points (size, std::numeric_limits<float>::max ());
+  
+  //set random seed
+  std::mt19937 random_gen(seed_);
+  std::uniform_int_distribution<index_t> dis(0, size -1);
+
+  //pick the first point at random
+  index_t max_index = dis(random_gen);
+  distances_to_selected_points[max_index] = -1.0;
+  indices.push_back(max_index);
+  
+  for (std::size_t j = 1; j < sample_size_; ++j)
+  {
+    index_t next_max_index = 0;
+    
+    //recompute distances
+    for (std::size_t i = 0; i < size; ++i)
+    {
+      if (distances_to_selected_points[i] != -1.0)
+        distances_to_selected_points[i] = std::min(distances_to_selected_points[i], geometry::distance((*input_)[i], (*input_)[max_index]));
```

Comment:
Very minor: Maybe lift the `(*input_)[max_index]` to a `const PointT& maxPoint = (*input_)[max_index]`. The two different points might be on very different cache lines, and the intent might become a bit more clear if a const Point reference prior is accessed.